headers/extract.py

```diff
@@ -118,76 +118,68 @@ def process_header(version, header):
     header = header.replace('\r\n', '\n')
     # remove some unused static const variables
     header = re.sub(r'(?m)^.*?const\s+.*?kIl2CppU?Int\w+(Min|Max)\s*=\s*.+\n', '', header)
-    header = re.sub(r'(?m)^.*?const\s+.*?kIl2CppNewLine.+\n', '', header)
+    header = re.sub(r'(?m)^.*?const\s+.*?kIl2Cpp(NewLine|SizeOf|OffsetOf).+\n', '', header)
     # remove useless pragmas
     header = re.sub(r'(?m)^\s*#pragma\s+once\s*\n', '', header)
     header = re.sub(r'(?m)^\s*#pragma\s+warning.+\n', '', header)
     # remove some C++-isms from older versions of the header (before they made it C compatible)
     header = re.sub(r'(?m)^template[^;]+;\n', '', header)
     header = re.sub(r'\bmutable\s+', '', header)
-    # remove blank lines
-    header = re.sub(r'(?m)^[ \t]*\n', '', header)
+    # remove C++ pure virtual function prototypes
+    header = re.sub(r'(?m)^\s*virtual .*? = 0;$', '', header)
 
     # IDA needs array sizes to be macros or literals, and doesn't understand const ints
     header = header.replace('[kPublicKeyByteLength]', '[8]')
     # Older versions used TypeInfo; change it to Il2CppClass for consistency with newer headers
     header = re.sub(r'\bTypeInfo\b', 'Il2CppClass', header)
     # Decompilation of array accesses only works properly when VLA arrays have nonzero length
-    header = header.replace('vtable[0]', 'vtable[32]')
+    header = re.sub(r'(?m)(\s+.*?)\s*\[0\]\s*;\s*$', r'\1[32];', header)
+    # Replace the nested il2cpp::os:: namespace definitions in 5.3.0-2017.4.40
+    header = re.sub(r'(?m)^namespace il2cpp\s*\{\s*namespace os\s*\{([^\}]+)\}\s*\}', r'\1', header)
+    # Replace all il2cpp::os:: usages
+    header = header.replace('il2cpp::os::', '')
+    # Remove Il2CppExceptionWrapper constructor
+    header = re.sub(r'(?m)^\s*Il2CppExceptionWrapper\s*\(Il2CppException\s*\*\s*ex\)\s*:\s*ex\s*\(ex\)\s*\{\}$', '', header)
+    # Remove const methods (GetIl2CppType and GetInternalThread
+    header = re.sub(r'(?m)^.*const$\s*\{[^\}]+\}', '', header)
+    # Remove exports defined with __attribute__ ((visibility ("default"))) (IL2CPPEXPORT/EXPORT)
+    # Basically iil2cpp_gc_wbarrier_set_field in current versions
+    header = re.sub(r'(?m)^.*?__attribute__ \(\(visibility \(\"default\"\)\)\).*$', '', header)
+    # Remove blank lines
+    header = re.sub(r'(?m)^[ \t]*\n', '', header)
+    # Change all class to struct
+    header = re.sub(r'(?m)^\s*class\s+', 'struct ', header)
+    # Change all C++ class inheritance to use composition for C compatibility
+    header = re.sub(r'(?m)^struct (.*?) : (?:public )?Il2Cpp(\S+)\s+\{([^\}]+)\};', r'struct \1 {\n Il2Cpp\2 \2;\3};', header)
```

Comment:
be careful here: MSVC and GCC use different techniques for laying out C++ inheritance, and the approach of just putting the base class as the first member only works for MSVC. Here's my notes on the matter:

- Il2Cpp does its own field offset computation in libil2cpp/metadata/FieldLayout.cpp, which is presumably used for generics. The basic layout algorithm is pretty bog-standard, but there's some subtleties.
  Given the declarations `struct A { int a; char b; }; struct B : A { char c; int d; char e; }; struct C : B { char f; };`, GCC and MSVC behave differently.

    - GCC lays C out as `struct C { struct A { int a; char b; } base; char c; int d; char e; char f; }` - that is, it encodes the root base class with normal alignment, then packs every other field inline.
    - MSVC lays C out as `struct C { struct B { struct A { int a; char b; } base; char c; int d; char e; } base; char f;};` - that is, it encodes all classes as if they contained their base classes as a separate struct, such that their alignments accumulate.

Reply:
In this example the fields all end up being laid out like a, b, c, d, e, f in the same order in memory anyway don't they, so does it matter?

Reply:
Due to structure alignment, the layouts would not exactly match up. Try `sizeof(struct C)` and `offsetof(struct C, f)` to see what I mean. On GCC, this will give 20 and 17; on MSVC, it will give 24 and 20.

Reply:
Proposed solution?